### PR TITLE
WT-12077 Fix zSeries checksum code

### PR DESCRIPTION
**Pull request by korteland**
See pinned comment on the ticket for a full description of the bug.

Not sure yet if this needs to go upstream - AFAICT both the "Linux on Z" IBM repo, and Linux itself are broken in the same way. I suspect they aren't checksumming small, misaligned values. And neither were we, until the new tests got merged.

(Adding our interns to the code review since this is kind of an interesting one.)

### CODE REVIEW

**Comment by korteland** [comment]
Added tests from all architectures to PR testing.

File: src/checksum/zseries/crc32-s390x.c [hunk 1]
```diff
@@ -92,18 +93,6 @@ __wt_checksum_hw(const void *chunk, size_t len)
 }
 #endif
 
-extern uint32_t __wt_checksum_sw(const void *chunk, size_t len);
```
Comment (sulabhM) [inline src/checksum/zseries/crc32-s390x.c:95]: Is it intentional to remove the externs

**Review by sulabhM — APPROVED** [review]

File: test/csuite/wt2695_checksum/main.c [hunk 2]
```diff
@@ -321,6 +324,17 @@ main(int argc, char *argv[])
         }
     }
 
+    /*
+     * "Strobed" misalignments - test every combo of size/misalignment up to 16B.
+     */
+    for (i = 0; i < 16; i++) {     /* Length. */
+        for (j = 0; j < 16; j++) { /* Misalignment. */
```
Comment (jeremythorp) [inline test/csuite/wt2695_checksum/main.c:331]: I'd suggest changing the names of i & j to length and misalignment and then the comments are no longer required.

**Review by jeremythorp — APPROVED** [review]
With one minor suggestion that LGTM!

File: test/csuite/wt2695_checksum/main.c (hunk 2 above)
Comment (korteland) [reply test/csuite/wt2695_checksum/main.c]: Thanks, you're right - I was being a bit lazy. Fixed in 7b663ae.

File: src/checksum/zseries/crc32-s390x.c (hunk 1 above)
Comment (korteland) [reply src/checksum/zseries/crc32-s390x.c:95]: Discussed offline, but for other reviewers - yes, we're now including `wiredtiger_internal.h` for `WT_MIN` so these are now duplicate declarations.

File: test/csuite/wt2695_checksum/main.c (hunk 2 above)
Comment (adelinexchen) [reply test/csuite/wt2695_checksum/main.c]: other people have also used i, j, k etc. (i.e. line 304 in the same file) is that bad? or is it ok in some circumstances

**Comment by adelinexchen** [comment]
lgtm :D

**Review by adelinexchen — APPROVED** [review]

File: test/csuite/wt2695_checksum/main.c (hunk 2 above)
Comment (korteland) [reply test/csuite/wt2695_checksum/main.c]: Good question @adelinexchen . I don't think there's a single yes/no answer, it's more of a matter of taste. For this new code, `i` and `j` represent fairly specific things (length and misalignment), so I took Jeremy's suggestion.

If you just need a generic counter, I reckon `i`/`j`/etc are fine.

Comment (adelinexchen) [reply test/csuite/wt2695_checksum/main.c]: @korteland Ah, thank you!!

File: test/csuite/wt2695_checksum/main.c [hunk 3]
```diff
@@ -321,6 +324,17 @@ main(int argc, char *argv[])
...
+     * "Strobed" misalignments - test every combo of size/misalignment up to 16B.
+     */
+    for (length = 0; length < 16; length++) {
+        for (misalignment = 0; misalignment < 16; misalignment++) {
+            hw = __wt_checksum(&data_ff[misalignment], length);
+            sw = __wt_checksum_sw(&data_ff[misalignment], length);
+            check(hw, sw, length, "0xff: strobed");
+        }
+    }
+
```
Comment (marcbutler) [inline test/csuite/wt2695_checksum/main.c:337]: Afaict this does not check that the tail length is correctly calculated as the length is never > VX_MIN_LEN.

Comment (korteland) [reply test/csuite/wt2695_checksum/main.c:337]: Yep, this is aimed at the head being misaligned. I can add a tail check if you feel it's valuable.

Comment (marcbutler) [reply test/csuite/wt2695_checksum/main.c:337]: Discussed in person. Additional effort does not seem warranted.

**Comment by marcbutler** [comment]
LGTM

**Review by marcbutler — APPROVED** [review]
LGTM

**Comment by korteland** [comment]
evergreen merge

**Pull request merged**